File: leetcode/1418.点菜展示表.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

/*
 * @lc app=leetcode.cn id=1418 lang=java
 *
 * [1418] 点菜展示表
 */

// @lc code=start
class Solution {
    /**
     *  分别统计每一个订单中的 桌号和菜品的名称
     */
    public List<List<String>> displayTable(List<List<String>> orders) {
        //菜品的名称Set Set会自动过滤相同的key值
        Set<String> nameSet = new HashSet<>();
        //每一卓的菜品统计 <桌号，<菜品名称，数量>>
        Map<Integer, Map<String, Integer>> foodsCnt = new HashMap<Integer, Map<String, Integer>>();
        for (List<String> order : orders) {
            nameSet.add(order.get(2));
            int id = Integer.parseInt(order.get(1));
            Map<String, Integer> map = foodsCnt.getOrDefault(id, new HashMap<String, Integer>());
            map.put(order.get(2), map.getOrDefault(order.get(2), 0) + 1);
            foodsCnt.put(id, map);
        }
        //因为要按照菜品的字母顺序升序 故进行排序，排序使用list
        List<String> nameList = new ArrayList<>();
        for(String name : nameSet){
            nameList.add(name);
        }
        //使用Collections工具类对nameList进行排序
        Collections.sort(nameList);

        //构建结果List
        List<List<String>> ans = new ArrayList<>();
        //第一行的数据 显示菜品
        List<String> first = new ArrayList<>();
        first.add("Table");
        for(int i = 0 ; i < nameList.size() ; i++){
            first.add(nameList.get(i));
        }
        //将第一行放入结果数组中
        ans.add(first);

        List<Integer> tables = new ArrayList<>();
        for(Integer i : foodsCnt.keySet()){
            tables.add(i);
        }
        //对桌号进行排序
        Collections.sort(tables);

        //遍历桌号并创建具体的餐品数量
        for(int i = 0 ; i < tables.size() ; i++){
            int id = tables.get(i);
            //获得对应桌号的菜品
            Map<String,Integer> cnt = foodsCnt.get(id);
            List<String> array = new ArrayList<>();
            array.add(String.valueOf(id));
            for(String name : nameList){
                array.add(String.valueOf(cnt.getOrDefault(name, 0)));
            }
            ans.add(array);
        }
        return ans;
    }
}
// @lc code=end

File: leetcode/golang/1418.点菜展示表.go
/*
 * @lc app=leetcode.cn id=1418 lang=golang
 *
 * [1418] 点菜展示表
 */

// @lc code=start
func displayTable(orders [][]string) [][]string {
	// 从订单中获取餐品名称和桌号，统计每桌点餐数量
    nameSet := map[string]struct{}{}
    foodsCnt := map[int]map[string]int{}
    for _, order := range orders {
		//strconv.Atoi stcconv是字符串转换包，Atoi是字符串转int的方法
        id, _ := strconv.Atoi(order[1])
        food := order[2]
        nameSet[food] = struct{}{}
		//统计餐品数量，如果当前桌号没有初始化则初始化
        if foodsCnt[id] == nil {
            foodsCnt[id] = map[string]int{}
        }
        foodsCnt[id][food]++
    }

	// 提取餐品名称，并按字母顺序排列  这里和java类似都需要用一个list类型
    n := len(nameSet)
    names := make([]string, 0, n)
    for name := range nameSet {
        names = append(names, name)
    }
	//sort包 是排序包Strings用于string排序
    sort.Strings(names)

	// 提取桌号，并按餐桌桌号升序排列
    m := len(foodsCnt)
    ids := make([]int, 0, m)
    for id := range foodsCnt {
        ids = append(ids, id)
    }
    sort.Ints(ids)

    // 填写点菜展示表
    table := make([][]string, m+1)
    table[0] = make([]string, 1, n+1)
    table[0][0] = "Table"
	//append 可以将一个数组的原色全部加入某一个数组 使用 ...即可
    table[0] = append(table[0], names...)
    for i, id := range ids {
        cnt := foodsCnt[id]
        table[i+1] = make([]string, n+1)
		//I to a 是int转为string类型的方法
        table[i+1][0] = strconv.Itoa(id)
        for j, name := range names {
            table[i+1][j+1] = strconv.Itoa(cnt[name])
        }
    }
    return table

}
// @lc code=end

